sematic/graph.py

```diff
@@ -0,0 +1,560 @@
+
+                if input_edge.source_run_id is not None:
+                    kwargs[input_edge.destination_name] = futures_by_original_id[
+                        input_edge.source_run_id
+                    ]
+                elif input_edge.artifact_id is not None:
+                    kwargs[input_edge.destination_name] = value
+                else:
+                    raise RuntimeError(
+                        "Invalid input edge had no source run or associated artifact"
+                    )
+
+            return kwargs, run_input_artifacts
+
+        def _get_run_output(run_id: RunID) -> Tuple[Any, Optional[Artifact]]:
+            output_edges = self._edges_by_source_id[run_id]
+
+            for output_edge in output_edges:
+                if output_edge.artifact_id is None:
+                    return None, None
+
+                artifact = self._artifacts_by_id[output_edge.artifact_id]
+                value = _get_artifact_value(artifact)
+                return value, artifact
+
+            return None, None
+
+        def _set_parent_nested(future: AbstractFuture, run: Run):
+            if run.parent_id is None:
+                return None, None
+
+            parent_future = futures_by_original_id[run.parent_id]
+
+            future.parent_future = parent_future
+
+            for output_edge in self._edges_by_source_id[run_id]:
+                if output_edge.parent_id is not None:
+                    if (
+                        self._edges_by_id[output_edge.parent_id].source_run_id
+                        == run.parent_id
+                    ):
+                        parent_future.nested_future = future
+
+                        if run.id in reset_run_ids:
+                            parent_future.state = FutureState.RAN
+
+                        return
+
+        def _clone_run(run_id: RunID):
+            run = self._runs_by_id[run_id]
+
+            kwargs, run_input_artifacts = _get_run_inputs(run_id)
+
+            func = run.get_func()
+
+            future = func(**kwargs)
+
+            input_artifacts[future.id] = run_input_artifacts
+
+            if run_id not in reset_run_ids:
+                value, output_artifact = _get_run_output(run_id)
+                if output_artifact is not None:
+                    output_artifacts[future.id] = output_artifact
+                    future.value = value
+
+            _set_parent_nested(future, run)
+
+            # Settings state for resolved runs unless reset
+            if FutureState[run.future_state] == FutureState.RESOLVED:  # type: ignore
+                if run.id not in reset_run_ids:
+                    future.state = FutureState.RESOLVED
+
+            futures_by_original_id[run.id] = future
+
+        for run_id in run_ids_by_execution_order:
+            if run_id in skip_run_ids:
+                continue
+
+            _clone_run(run_id)
+
+        if reset_from is None and len(futures_by_original_id) != len(self.runs):
```

Comment:
Why is this check only valid when `reset_from` is None?

Reply:
Because otherwise certain runs are skipped. E.g. descendants of `reset_from`, descendants of downstreams of ancestors of `reset_from`.